active_discussion/embedded/critical-section.md

```diff
@@ -0,0 +1,290 @@
+
+Note that "any safe code" can *not* call `main` or `INTERRUPT0` (because they
+are `unsafe` functions), use `asm!` or access registers.
+
+**Claim**:  This program is well-defined / sound if and only if `Type`
+implements the `Send` trait.
+
+Example that shows that the bound is required: `type Type = Rc<u8>` could
+result in an unsound program (data race between `main` and `INTERRUPT0`).
+
+"Why are the memory clobbers required?" Without them the compiler can reorder
+`main`'s operations on `X` to outside the critical section leading to a data
+race.
+
+## Interrupt masking
+
+Consider this program that creates a critical section by masking a single
+interrupt (individual mask).
+
+> Aside: it's also possible to implement a critical section by raising the
+> running priority but the implementation of that kind of critical section is
+> very similar to this one (volatile write + compiler fence) so we won't include
+> it in this document.
+
+``` rust
+#![no_std]
+
+use core::{cell::UnsafeCell, ptr, sync::atomic::{self, Ordering}};
+
+extern "C" {
+    static MASK_INTERRUPT: UnsafeCell<u8>;
+    static UNMASK_INTERRUPT: UnsafeCell<u8>;
+}
+
+static mut X: Type = Type::default();
+
+#[no_mangle]
+unsafe fn main() -> ! {
+    unsafe {
+        asm!("ENABLE_INTERRUPTS" : : : : "volatile");
+    }
+
+    loop {
+        // .. any safe code ..
+
+        unsafe {
+            // start of critical section
+            ptr::write_volatile(MASK_INTERRUPT.get(), 1 << 0);
+        }
+
+        atomic::compiler_fence(Ordering::SeqCst);
+
+        // `INTERRUPT0` can *not* preempt this block
+        // (because it's masked)
+        {
+            let x: &mut Type = unsafe {
+                &mut X
+            };
+
+            // .. any safe code ..
+        }
+
+        atomic::compiler_fence(Ordering::SeqCst);
+
+        unsafe {
+            // end of critical section
+            ptr::write_volatile(UNMASK_INTERRUPT.get(), 1 << 0);
+        }
+
+        // .. any safe code ..
+    }
+}
+
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    let x: &mut Type = unsafe {
+        &mut X
+    };
+
+    // .. any safe code ..
+}
+```
+
+**Claim**:  This program is well-defined / sound if and only if `Type`
+implements the `Send` trait.
+
+Example that shows that the bound is required: `type Type = Rc<u8>` could
+result in an unsound program (data race between `main` and `INTERRUPT0`).
+
+"Why are the compiler fences required?" Without them the compiler can reorder
+`main`'s operations on `X` to outside the critical section leading to a data
+race.
+
+## Questions
+
+- Can these programs be misoptimized by the compiler? In particular, the
+  compiler fences in the second program prevent memory operations on `X` from
+  being reordered to outside the critical section but AFAIK they don't tell the
+  compiler that `X` may change outside the critical section -- could the program
+  cache the value of `X` on the stack? That would change the semantics of the
+  program.
```

Comment:
Actually that is the whole point of a compiler fence: it tells the compiler that other threads may have modified shared data (i.e. globals + any locals whose address has been made available globally) and that it should reload those at the point of the fence.

Reply:
It also flushes writes made before the fence, which is why `memory` is required on the second `asm!` (that re-enables interrupts).

Reply:
@Amanieu is that documented somewhere? That's not what I understood from the
[`atomic::compiler_fence`] documentation; that doc only talks about preventing
the compiler from reordering memory operations across the fence; it says
nothing about forcing shared / static variables to be reloaded.

[`atomic::compiler_fence`]: https://doc.rust-lang.org/core/sync/atomic/fn.compiler_fence.html

I can certainly see the behavior you mention arising from `asm!("" : : :
"memory")` where the compiler has to assume that the assembly block may modify
any memory, but the text you have commented on is asking about the program that
exclusively uses `atomic::compiler_fence`. Or do you mean to say that I should
consider `asm!("" : : : "memory")` and `atomic::compiler_fence` to be
equivalent? IME, they do have observable differences on optimizations.

Reply:
If the source code loads a variable twice but the emitted assembly only loads it once, the compiler has effectively reordered the second load up to the same point as the first load.

`atomic::compiler_fence` emits a LLVM [`fence` instruction](https://llvm.org/docs/LangRef.html#fence-instruction) with `syncscope` set to `singlethread`.  As mentioned in a different thread, it's also used for C `atomic_signal_fence`, which is made specifically for the interrupts use case:

> [Establishes memory synchronization ordering of non-atomic and relaxed atomic accesses, as instructed by order, between a thread and a signal handler executed on the same thread.](https://en.cppreference.com/w/cpp/atomic/atomic_signal_fence)

(It says "signal handler", but from the compiler's perspective, a Unix signal handler in userland and an interrupt handler on bare metal have basically the same semantics; they both interrupt execution at an arbitrary instruction and move control flow to somewhere else.)

I think `asm!("" : : : "memory")` should be largely if not entirely equivalent to `atomic::compiler_fence` with either `AcqRel` or `SeqCst` (as opposed to `Acquire` and `Release`, which only create one-sided barriers).  I'd be interested to see what code creates differences in assembly output between the two.  Since they are different LLVM constructs, it's possible that LLVM's optimizer could just coincidentally generate different code (in particular, it might be being overly conservative with `asm!`), but it's also possible that I'm forgetting something.

Reply:
@japaric 

> That's not what I understood from the
atomic::compiler_fence documentation; that doc only talks about preventing
the compiler from reordering memory operations across the fence; it says
nothing about forcing shared / static variables to be reloaded.

The RFC that added these describes their semantics as equivalent to C11's `atomic_signal_fence`. The current docs do not make that clear, but I'd argue that this is a documentation bug.

Reply:
@comex

> If the source code loads a variable twice but the emitted assembly only loads
> it once, the compiler has effectively reordered the second load up to the same
> point as the first load.

That clarifies things; thanks.

(Thinking of #112) Is removing a store operation considered a re-ordering?

> I'd be interested to see what code creates differences in assembly output
> between the two

`compiler_fence` prevents memory operations on stack variables from being
merged / reordered. The "memory" clobber seems to have no effect on stack
variables. See code below:

``` rust
fn SysTick() {
    let mut x = 0;

    atomic::compiler_fence(Ordering::AcqRel);

    x += 1;

    unsafe {
        // prevent LLVM from optimizing away `x`
        ptr::read_volatile(&x);
    }
}
```

``` armasm
SysTick:
        sub     sp, #4
        movs    r0, #0
        str     r0, [sp] ; x = 0
        ldr     r0, [sp]
        adds    r0, #1
        str     r0, [sp] ; x += 1
        ldr     r0, [sp] ; read_volatile(&x)
        add     sp, #4
        bx      lr
```

vs

``` rust
fn SysTick() {
    let mut x = 0;

    unsafe { asm!("" : : : "memory" : "volatile") }

    x += 1;

    unsafe {
        // prevent LLVM from optimizing away `x`
        ptr::read_volatile(&x);
    }
}
```

``` armasm
SysTick:
        sub     sp, #4
        movs    r0, #1
        str     r0, [sp] ; x = 1
        ldr     r0, [sp] ; read_volatile(&x)
        add     sp, #4
        bx      lr
```

If you replace the stack variable with a static variable then both the compiler
fence and the "memory" clobber prevent merging / re-ordering of memory
operations on the static variable.

Reply:
> (Thinking of #112) Is removing a store operation considered a re-ordering?

Not sure what you are asking? It is not, but once you reordered enough to get two adjacent stores you can obviously remove the first.

Reply:
> compiler_fence prevents memory operations on stack variables from being
merged / reordered. The "memory" clobber seems to have no effect on stack
variables. See code below:

To me this looks like an arbitrary LLVM limitation and not a fundamental difference. All clobber and orders aside, the compiler can in both cases assume that `x` is unobserved by the outside world because its address is not leaked to anything. So IMO in both cases it would be legal to optimize away all the writes, and just leave the volatile read.